runtime/autoload/man.vim

```diff
@@ -11,127 +20,197 @@ catch /E145:/
   " Ignore the error in restricted mode
 endtry
 
-" Load man page {page} from {section}
-"   call man#get_page([{section}, ]{page})
-function man#get_page(...) abort
-  let invoked_from_man = (&filetype ==# 'man')
-
-  if a:0 == 0
-    echoerr 'argument required'
+function! man#get_page(count, editcmd, ...) abort
+  if a:0 > 2
+    call s:error('too many arguments')
     return
-  elseif a:0 > 2
-    echoerr 'too many arguments'
+  elseif a:0 == 0
+    call s:error('what manual page do you want?')
     return
+  elseif a:0 == 1
+    let [page, sect] = s:parse_page_and_sect_fpage(a:000[0])
+    if (empty(sect) && sect != 0) && a:count != 10
+      let sect = a:count
+    endif
+  else
+    let sect = tolower(a:000[0])
+    let page = a:000[1]
   endif
 
-  let sect = get(a:000, 0)
-  let page = get(a:000, 1, sect)
+  let out = systemlist(s:man_cmd.s:man_find_arg.' '.s:man_args(sect, page))
+  if empty(out) || out[0] == ''
+    call s:error('no manual entry for '.page.(s:empty_sect(sect)?'':'('.sect.')'))
+    return
+  elseif page !~# '\/' " if page is not a path, parse the page and section from the path
+    " use the last line because if we had something like printf(man) then man
+    " would be read as the manpage because man's path is at out[0]
+    let [page, sect] = s:parse_page_and_sect_path(out[len(out)-1])
+  endif
 
-  let [page, sect] = s:parse_page_and_section(sect, page)
+  call s:push_tag()
 
-  if !empty(sect) && s:find_page(sect, page) == 0
-    let sect = ''
-  endif
+  call s:read_page(sect, page, a:editcmd)
+endfunction
 
-  if s:find_page(sect, page) == 0
-    echo 'No manual entry for '.page
-    return
+" move to previous position in the stack
+function! man#pop_tag() abort
+  if !empty(s:tag_stack)
+    let tag = remove(s:tag_stack, -1)
+    execute tag['buf'].'b'
+    call cursor(tag['lnum'], tag['col'])
   endif
+endfunction
 
-  exec 'let s:man_tag_buf_'.s:man_tag_depth.' = '.bufnr('%')
-  exec 'let s:man_tag_lin_'.s:man_tag_depth.' = '.line('.')
-  exec 'let s:man_tag_col_'.s:man_tag_depth.' = '.col('.')
-  let s:man_tag_depth = s:man_tag_depth + 1
-
-  let editcmd = 'edit'
-  " Use an existing 'man' window, else open a new one.
-  if &filetype !=# 'man'
-    let thiswin = winnr()
-    wincmd b
-    if winnr() > 1
-      exec thiswin . 'wincmd w'
-      while 1
-        if &filetype ==# 'man'
-          break
-        endif
-        wincmd w
-        if thiswin == winnr()
-          break
-        endif
-      endwhile
-    endif
+" save current position in the stack
+function! s:push_tag() abort
+  let s:tag_stack += [{
+        \ 'buf':  bufnr('%'),
+        \ 'lnum': line('.'),
+        \ 'col':  col('.'),
+        \ }]
+endfunction
 
-    if &filetype !=# 'man'
-      let editcmd = 'tabnew'
+" find the closest man window above/left
+" TODO(nhooyr) more ways to open manpages
```

Comment:
Need to figure out a way to also allow opening a manpage in the current window/vertically/horizontally.